### ISSUE TITLE
Run Hasura backend locally

### ISSUE THREAD

**Issue by SamLubbers** [xrnl/aginun#175]
## Problem that this feature solves

This feature will make the installation of the application much easier, will enable other XR groups to run an instance of the application, and will save us money.

The code for the front-end of the is all contained in this repository and can be installed & run with the commands in the readme. This is very useful because every person can run their own instance of the application. However, we only have one (Hasura) back-end, which is running on a Virtual Private Server and is configured manually. This is a problem because:

1. For every new instance of the application that we want to run we need to go through the laborious and error-prone process of setting up Hasura, connecting it to the front-end etc.
2. Other XR groups who want to use their own version of the volunteer platform cannot do so because all front-ends connect to our shared development backend.
3. We need to pay every month for the VPS that we use for the development backend.

## Feature description

What I propose is incorporate the back-end into this code base, so with a single command, you run both the backend and front-end, and these automatically connect to each other via a local network on the same machine. Achieving this requires different steps:

### 1. Create docker-compose

With [docker-compose](https://docs.docker.com/compose/) it is possible to start and stop multiple services with a single command. Our goal is to have a single docker-compose file that starts both the frontend (current code in the repository) and backend (Hasura, currently running in the VPS). Luckily, Hasura is already running with docker-compose, so all we need to do is to:

1. Get the existing Hasura docker-compose file
2. Adapt the docker-compose with an additional service for the front-end
3. Configure the front-end to connect to the backend specified in the docker-compose. The line `uri: "http://178.62.229.109/v1/graphql"` in `src/plugins/vue-apollo.ts` should be changed to the IP address used by the local Hasura backend. Of course, the IP address of the Hasura backend should be exposed by docker so that the front-end can connect to it.

### 2. Populate Back-end database

After following these steps, you should have a running front-end and a working back-end. However, when you try loading the front-end it will not work because there is still no data in the Hasura database. Therefore, the PostgreSQL database used by Hasura must have:

1. All the tables & functions configured in the backend. Access Hasura through [the console](http://178.62.229.109/console/data/default) to see all defined tables & functions. We could achieve this by getting a `pg_dump` of the existing database used by Hasura, and then adding a service to the `docker-compose` which creates the database tables on the first load.
2. The `workingCircle` and `localGroup` tables should be populated, so that users can select those groups when creating new roles, and when they filter existing roles. We cannot load this from the docker-compose because XR groups in other countries will have different local groups and working groups. It would be better if we have admin users that can define these groups in a dedicated admin view. Perhaps we should get this working first with the existing backend, and then do all other steps. (see #180)

### 3. Different configuration for development & deployment

When we're running the application in development we want a different setup than in production:

* In development we want the front-end to run with the command `npm start`, whereas in production we want to run `npm run build` and serve the production app on the right place of our deployment server.
* In production we need to enable HTTPs when accessing the front-end, or the console of the Hasura backend. 

### 4. Update readme

Once the whole new setup is working, we need to update the readme so that it has the correct instructions for running the application.

## Additional comments

If it's too difficult to configure everything in one `docker-compose` we could also first start with just ensuring that we can run the backend locally. Starting the application would then require having two separate process that we need to start manually. Then we can make another issue for grouping everything in one go.

If you are going to work on this please [contact us](https://github.com/xrnl/aginun/blob/master/SUPPORT.md#get-in-contact) to get access to the current configuration of the Hasura backend server.

**Comment by SamLubbers** [comment]
FInally, once this is merged & working we should get rid of the VPS that we use for the development backend server so that we don't have to keep for it every month.

**Comment by SamLubbers** [comment]
In the meeting of [15/06/2021](https://cloud.extinctionrebellion.nl/index.php/s/dHzKmXNzQoqrcCS) Ale mentioned an important new requirement for this task: if the backend stops working, the website should still serve a page with a friendly error message telling the user to contact us for support. If we don't have this, then users won't even be able to access the page if there are problems in the backend.

I think this will already work fine with our current setup with separate docker containers. But we should test it just in case.

**Comment by jurcello** [comment]
> In the meeting of [15/06/2021](https://cloud.extinctionrebellion.nl/index.php/s/dHzKmXNzQoqrcCS) Ale mentioned an important new requirement for this task: if the backend stops working, the website should still serve a page with a friendly error message telling the user to contact us for support. If we don't have this, then users won't even be able to access the page if there are problems in the backend.
> 
> I think this will already work fine with our current setup with separate docker containers. But we should test it just in case.

I would create a new issue for this so we can keep the tasks small.
